src/LanguageServer/Impl/Sources/CodeActionSource.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Python.Analysis;
+using Microsoft.Python.Analysis.Analyzer;
+using Microsoft.Python.Analysis.Diagnostics;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.Collections;
+using Microsoft.Python.LanguageServer.CodeActions;
+using Microsoft.Python.LanguageServer.Protocol;
+using Microsoft.Python.Parsing.Ast;
+using Range = Microsoft.Python.Core.Text.Range;
+
+namespace Microsoft.Python.LanguageServer.Sources {
+    internal sealed partial class CodeActionSource {
+        private static readonly ImmutableArray<ICodeActionProvider> _codeActionProviders =
+            ImmutableArray<ICodeActionProvider>.Create(MissingImportCodeActionProvider.Instance);
+
+        private readonly IServiceContainer _services;
+
+        public CodeActionSource(IServiceContainer services) {
+            _services = services;
+        }
+
+        public async Task<CodeAction[]> GetCodeActionsAsync(IDocumentAnalysis analysis, Range range, Diagnostic[] diagnostics, CancellationToken cancellationToken) {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var results = new List<CodeAction>();
+
+            // * NOTE * this always re-calculate whole document for all linters.
+            //          consider creating linter result cache somewhere (MRU?) and 
+            //          add error code as optional argument to run only that linter rather
+            //          than always running all linters.
+            //          also, current LintModule implementation always run all linters
+            //          even if the linter option is off to set references in the given
+            //          modules. this code path should use different code path to prevent that.
```

Comment:
Yes, running all is currently by design as it is part of the analysis and we don't want to re-run that when option changes. So user options are basically display options. We try and separate linter pass but some 'linting' messages are actually produced by the analysis run.

Reply:
I meant running Linter for code action.

Reply:
You don't need to run it. It was ran during analysis that you are using and any issues are in the `IDiagnosticService`.